packages/@expo/dev-server/src/LaunchBrowser.ts

```diff
@@ -0,0 +1,268 @@
+import * as osascript from '@expo/osascript';
+import spawnAsync from '@expo/spawn-async';
+import { spawn, type ChildProcess } from 'child_process';
+import { sync as globSync } from 'glob';
+import open from 'open';
+import os from 'os';
+import path from 'path';
+
+const IS_WSL = require('is-wsl') && !require('is-docker')();
+
+export interface BrowserInstance {
+  close: () => Promise<void>;
+}
+
+/**
+ * Supported browser types
+ */
+enum BrowserTypes {
+  CHROME,
+  EDGE,
+}
+
+/**
+ * Internal browser implementation constraints
+ */
+interface BrowserImpl {
+  /**
+   * Return whether the given `browserType` is supported
+   */
+  isSupportedBrowser: (browserType: BrowserTypes) => Promise<boolean>;
+
+  /**
+   * Launch the browser
+   */
+  launchAsync: (browserType: BrowserTypes, args: string[]) => Promise<BrowserInstance>;
+
+  /**
+   * Close current browser instance
+   */
+  close: () => Promise<void>;
+}
+
+/**
+ * Launch a browser for JavaScript inspector
+ */
+export async function launchBrowserAsync(url: string): Promise<BrowserInstance> {
+  // For dev-client connecting metro in LAN, the request to fetch sourcemaps may be blocked by Chromium
+  // with insecure-content (https page send xhr for http resource).
+  // Adding `--allow-running-insecure-content` to overcome this limitation
+  // without users manually allow insecure-content in site settings.
+  // However, if there is existing chromium browser process, the argument will not take effect.
+  // We also pass a `--user-data-dir=` as temporary profile and force chromium to create new browser process.
+  const tempBrowserDir = await createTempBrowserDir();
+  const launchArgs = [
+    `--app=${url}`,
+    '--allow-running-insecure-content',
+    `--user-data-dir=${tempBrowserDir}`,
+    '--no-first-run',
+    '--no-default-browser-check',
+  ];
+
+  const browser = createBrowser();
+
+  for (const browserType of [BrowserTypes.CHROME, BrowserTypes.EDGE]) {
+    const isSupported = await browser.isSupportedBrowser(browserType);
+    if (isSupported) {
+      return browser.launchAsync(browserType, launchArgs);
+    }
+  }
+
+  throw new Error(
+    '[LaunchBrowser] Unable to find a browser on the host to open the inspector. Supported browsers: Google Chrome, Microsoft Edge'
+  );
+}
+
+function createBrowser() {
+  if (os.platform() === 'darwin') {
+    return new BrowserImplMacOS();
+  }
+  if (os.platform() === 'win32' || IS_WSL) {
+    return new BrowserImplWindows();
+  }
+  if (os.platform() === 'linux') {
+    return new BrowserImplLinux();
+  }
+  throw new Error('[LaunchBrowser] Unsupported host platform');
+}
+
+/**
+ * Create a temp folder for chromium user profile
+ */
+async function createTempBrowserDir() {
```

Comment:
Would be good to have a test for this function.

Reply:
that is more a platform thin wrapper. if we don't have windows or wsl testing runners, it doesn't have much benefit from this test this function.
i'm moving this function to LaunchBrowserImpl as a platform-dependent implementation.